### PR TITLE
Add shared memories

### PR DESCRIPTION
**Pull request by abrown**
This change adds the ability to use shared memories in Wasmtime when the
[threads proposal] is enabled. Shared memories are annotated as `shared`
in the WebAssembly syntax, e.g., `(memory 1 1 shared)`, and are
protected from concurrent access during `memory.size` and `memory.grow`.

[threads proposal]: https://github.com/WebAssembly/threads/blob/master/proposals/threads/Overview.md

In order to implement this in Wasmtime, there are two main cases to
cover:
    - a program may simply create a shared memory and possibly export it;
    this means that Wasmtime itself must be able to create shared
    memories
    - a user may create a shared memory externally and pass it in as an
    import during instantiation; this is the case when the program
    contains code like `(import "env" "memory" (memory 1 1
    shared))`--this case is handled by a new Wasmtime API
    type--`SharedMemory`

Because of the first case, this change allows any of the current
memory-creation mechanisms to work as-is. Wasmtime can still create
either static or dynamic memories in either on-demand or pooling modes,
and any of these memories can be considered shared. When shared, the
`Memory` runtime container will lock appropriately during `memory.size`
and `memory.grow` operations; since all memories use this container, it
is an ideal place for implementing the locking once and once only.

The second case is covered by the new `SharedMemory` structure. It uses
the same `Mmap` allocation under the hood as non-shared memories, but
allows the user to perform the allocation externally to Wasmtime and
share the memory across threads (via an `Arc`). The pointer address to
the actual memory is carefully wired through and owned by the
`SharedMemory` structure itself. This means that there are differing
views of where to access the pointer (i.e., `VMMemoryDefinition`): for
owned memories (the default), the `VMMemoryDefinition` is stored
directly by the `VMContext`; in the `SharedMemory` case, however, this
`VMContext` must point to this separate structure.

To ensure that the `VMContext` can always point to the correct
`VMMemoryDefinition`, this change alters the `VMContext` structure.
Since a `SharedMemory` owns its own `VMMemoryDefinition`, the
`defined_memories` table in the `VMContext` becomes a sequence of
pointers--in the shared memory case, they point to the
`VMMemoryDefinition` owned by the `SharedMemory` and in the owned memory
case (i.e., not shared) they point to `VMMemoryDefinition`s stored in a
new table, `owned_memories`.

This change adds an additional indirection (through the `*mut
VMMemoryDefinition` pointer) that could add overhead. Using an imported
memory as a proxy, we measured a 1-3% overhead of this approach on the
`pulldown-cmark` benchmark. To avoid this, Cranelift-generated code will
special-case the owned memory access (i.e., load a pointer directly to
the `owned_memories` entry) for `memory.size` so that only
shared memories (and imported memories, as before) incur the indirection
cost.

### CODE REVIEW

File: crates/wasmtime/src/memory.rs
Comment (abrown) [inline crates/wasmtime/src/memory.rs:817]: I think a refactoring to make this work would be to:
 - change the signature of `RuntimeLinearMemory::grow` to take a `limiter_fn: Option<Fn<...>>` instead of the entire store
 - remove the error reporting to the limiter (`Store::memory_grow_failed` -> `ResourceLimiter::memory_grow_failed`); right now it is unclear (except during testing) why a limiter would need to receive the growth failure--shouldn't it be the one producing this error instead? If indeed this _is_ needed, perhaps we could alter the suggestion above to an `Option<dyn ResourceLimiter>`?

File: crates/cranelift/src/compiler.rs
Comment (abrown) [inline crates/cranelift/src/compiler.rs:335]: It is unclear what to do at this point if the module's memory is a shared one... The current logic below is incorrect.

Comment (alexcrichton) [reply crates/cranelift/src/compiler.rs:335]: Given that this is only debuginfo related and debuginfo is off-by-default I think it'd be fine to just return an error here for shared memories and have someone else come along later and figure this out if necessary.

File: crates/environ/src/module.rs
Comment (alexcrichton) [reply crates/environ/src/module.rs:937]: (only applicable for multi-memory of course, but still something we'll want to handle, and perhaps a good test case to add eventually as well)

File: crates/environ/src/vmoffsets.rs
Comment (alexcrichton) [inline crates/environ/src/vmoffsets.rs:167]: Similar to above, I think that the calculation here needs to exclude imported memories first, and then exclude shared memories from the remaining set of memories to figure out the number of owned memories.

File: crates/runtime/src/instance.rs
Comment (alexcrichton) [inline crates/runtime/src/instance.rs:214]: This I think has pretty deep implications which require an audit of callers of this function to ensure they still work correctly. Looking I see:

* `Instance::memory` - this is not ok for shared memories since it does a non-atomic read of the pointer
* `Instance::set_memory` - this is no longer valid and needs to be updated. Callers of this function are:
  * `Instance::memory_grow` - this'll need to be updated since the update only happens for owned memories
* `Instance::get_exported_memory` - this is ok but I think callers of this need to be updated to do atomic reads of the length field
* `Instance::memory_index` - this is no longer valid and needs to be updated (or deleted)

File: crates/runtime/src/memory.rs
Comment (alexcrichton) [inline crates/runtime/src/memory.rs:174]: IIRC the reason for passing them explicitly was that it's somewhat tricky to calculate the maximum/minimum converting from u32-or-u64 to `usize` so by passing them here it's "pre-validated" from the results of the previous calculations.

File: crates/runtime/src/memory.rs
Comment (alexcrichton) [inline crates/runtime/src/memory.rs:497]: Since the outer `SharedMemory` is boxed up as well, could this perhaps be `SharedMemoryInner<T>` and the memory here is `memory: RwLock<T>`?

File: crates/runtime/src/memory.rs
Comment (alexcrichton) [inline crates/runtime/src/memory.rs:522]: This growth operation, on success, needs to perform an atomic increment of the `LongTermVMMemoryDefinition` current length I think?

File: crates/runtime/src/memory.rs
Comment (alexcrichton) [inline crates/runtime/src/memory.rs:646]: This change is actually a bit worrisome to me because the purpose of this argument is to limit memory usage within a `Store`, but our current strategy of a per-`Store` memory limiter is not sufficient to protect against shared memories.

If `store` is `None` could this assert that the memory is shared? Additionally I think we'll need to keep a note on a tracking issue for shared memories to figure out how best to limit them since our current `ResourceLimiter` won't work

File: crates/runtime/src/memory.rs
Comment (alexcrichton) [inline crates/runtime/src/memory.rs:453]: Could this function take `MemoryPlan` instead of recalculating it here?

Also, could this internally validate that the plan is `Static`?

File: crates/runtime/src/memory.rs
Comment (alexcrichton) [reply crates/runtime/src/memory.rs:497]: Or actually I see now that the downcast into a shared memory might be difficult with that, so could this otherwise be hardcoded as an `MmapMemory`?

File: crates/wasmtime/src/memory.rs
Comment (alexcrichton) [reply crates/wasmtime/src/memory.rs:817]: Oh sorry I sort of independently came to a similar conclusion that this would be a problem above before I saw your comment here.

I think for now let's skip limiting memory growth of shared memories with `ResourceLimiter` so this would thread through a `None` or similar (and wherever uses the limiter would assert the store is present if the memory isn't shared).

Overall we're going to need a different strategy for limiting shared memories since growth would notify a random store which isn't necessarily the right "store". I would naively say that we'll need a limiter-per-shared-linear-memory but we can figure that out later, no need to have it covered for now.

**Review by alexcrichton — COMMENTED** [review]
I think this is some good progress! I realize this might already be getting to be somewhat of an unwieldy PR so if you'd like I think various things below can also be deferred to later. If you'd prefer to do that I think it would be good to make an issue with remaining work items and to make sure everything is reflected in that before landing this.

One other issue not mentioned below though that I want to be sure to not forget is that all of Wasmtime's interactions with shared memories through intrinsics like `memory.init` are going to need to be updated for shared memories. I think in general Wasmtime needs to be very carefully audited for the rampant usage of `VMMemoryDefinition` and have that all locked down through some different abstraction in the case of shared or non-shared memory. I think a static distinction at the top level with `Memory` and `SharedMemory` will help with this (I commented below but I think we may actually want to prevent a `SharedMemory` from being viewed as a `Memory`) but we'll still want to audit internals for implementations like `memory.init`, `memory.fill`, `memory.copy`, etc.

File: crates/runtime/src/memory.rs
Comment (abrown) [reply crates/runtime/src/memory.rs:547]: Done in `grow` instead.

File: crates/environ/src/module.rs
Comment (abrown) [reply crates/environ/src/module.rs:928]: `defined_memory_index` a few lines above was returning `Option<...>` so I kept the same level of abstraction--`defined_memory_index` can fail by indexing into the imported memories and this one can fail by not finding an owned memory within the index space. Looking at it again, now I'm wondering if the `.skip(...)` actually makes the lower part of this incorrect...

File: crates/runtime/src/memory.rs
Comment (cfallin) [inline crates/runtime/src/memory.rs:546]: I'm not sure I fully understand the atomicity guarantees around how the actual growth (making pages accessible) and setting the new size (below) occur. A comment here might be good to have, explaining why it is OK for these to be two separate steps, and what happens if other concurrently-running threads run Wasm that use `memory.size` and probe pages for accessibility?

Specifically I'm curious about the case where another thread (i) accesses a memory address, finding that it does not trap, and execution continues; (ii) runs `memory.size`, and sees that the memory is supposedly smaller than the just-accessed address. This is possible if one imagines this (growing) thread to pause between the `grow()` above and the atomic store below.

Do we say that this is a racy Wasm program and so results are undefined? In other words, are all memory accesses that implicitly depend on the memory length actually non-synchronizing, and any Wasm program that depends on a certain access to an address trapping must ensure it synchronizes with any other threads growing the shared memory?

That seems like a reasonable enough answer to me, but we should (i) document it, and (ii) confirm that it's spec-compliant. And actually I hope that it is our answer, because providing a stronger unit of atomicity seems really hard (pause the world? Or set the new length first, and catch and handle a false out-of-bounds trap by taking a lock on length and re-checking to be sure in the signal handler?).

File: crates/runtime/src/vmcontext.rs
Comment (cfallin) [inline crates/runtime/src/vmcontext.rs:231]: This doc-comment seems a little unclear to me: if the result isn't exactly precise at all times, how might it differ? Is it always smaller or equal (because memory can only grow but it may be outdated)? Or something else? It seems we need a stronger statement here to rely on it at all.

I also wonder if this is necessary vs. a variant that uses `Ordering::Acquire` and makes a stronger claim? Or in other words, can we justify why a "fast and imprecise" variant is needed?

File: crates/cranelift/src/func_environ.rs
Comment (cfallin) [inline crates/cranelift/src/func_environ.rs:1732]: A comment here noting that this is the only access to the length field from JIT'd code for a shared memory would be helpful. Specifically I was worried that a dynamically-bounds-checked access could exist and went looking for logic that would generate atomic loads in the `heap_addr` lowering, or some equivalent change; then realized that we prevent that case by construction, by only allowing `Static`-type shared memories. Could we document that here?

File: crates/cranelift/src/func_environ.rs
Comment (cfallin) [reply crates/cranelift/src/func_environ.rs:1381]: I think in the sense of "make sure we don't feed `current_length` to the heap-access machinery in Cranelift"-safety, we could maybe make `current_length_offset` an `Option` here, and return `None` for the shared case. Then below wherever we use it, unwrap it with "expected current-length field (always present on non-shared memories)" or something of the sort as an `expect` error?

**Review by cfallin — COMMENTED** [review]
I made a general pass over this and added a few questions below. I'll defer to @alexcrichton for final approval, but some of the things below I think would help to clarify and boost confidence a bit. Thanks!

File: crates/runtime/src/vmcontext.rs
Comment (alexcrichton) [reply crates/runtime/src/vmcontext.rs:231]: Actually that raises another question for me about atomic orderings which sort of furthers my concerns about using `ptr::copy` for `memory.copy` and friends. If one thread witnesses that the shared memory has a particular size I don't know what the guarantees are about witnessing other parts of memory in terms of memory orderings. For example should this be `Acquire`? Should this be `SeqCst`? Do wasm modules otherwise need to `atomic.fence` before calling host functionality or something like that?

I personally know very little about the atomic memory model for wasm and how it's all intended to work out so there may not be a great answer here. That being said it's almost certainly not wrong to fall back to `SeqCst` so I might say we should do that to be maximally conservative?

File: crates/runtime/src/vmcontext.rs
Comment (abrown) [reply crates/runtime/src/vmcontext.rs:136]: There's no `vmmemory_import_index` function to test against... should there be? (I don't think we use the index offset in the JIT-generated code).

File: crates/wasmtime/src/memory.rs
Comment (abrown) [reply crates/wasmtime/src/memory.rs:717]: Looking at `Tunables` (now for the n-th time), I can't really see how we could guarantee that only a static allocation is used.

File: crates/cranelift/src/func_environ.rs
Comment (abrown) [reply crates/cranelift/src/func_environ.rs:1738]: Welp, not any more after removing the `Option` from the return of `owned_memory_index`!

File: crates/wasmtime/src/memory.rs
Comment (alexcrichton) [reply crates/wasmtime/src/memory.rs:265]: The `vmimport` constructor doesn't need to create a `Memory` through which inserts into the memories table within the `Store`, so I think that the call to the trampoline helper here can be inlined into `vmimport` and then the instance handle can be used to extract the memory directly rather than going through `Memory`

File: crates/wasmtime/src/memory.rs
Comment (alexcrichton) [reply crates/wasmtime/src/memory.rs:718]: I don't think you'll want to determine that by looking at the tunables but rather by looking at the `MemoryPlan` that pops out. Additionally though this isn't the best place for the error but rather in `wasmtime_runtime::SharedMemory` during the constructor you can `match` on the `MemoryPlan` and reject `Dynamic` variants.

File: crates/runtime/src/memory.rs
Comment (abrown) [reply crates/runtime/src/memory.rs:563]: I tried but we aren't there yet: `instance.rs` uses this function both to setup the `vmctx` initially in `initialize_vmctx` and later to replace the `VMMemoryDefinition` as a part of `memory_grow`.

Comment (alexcrichton) [reply crates/runtime/src/memory.rs:563]: Isn't that a bug though that needs fixing? The write of `VMMemoryDefinition` isn't atomic so it needs to be synchronized with everything else. I think that `memory_grow` needs to be updated to only update the local owned context and otherwise rely on the receiver to update?

File: crates/runtime/src/memory.rs
Comment (abrown) [reply crates/runtime/src/memory.rs:123]: https://github.com/bytecodealliance/wasmtime/issues/4240

File: crates/runtime/src/memory.rs
Comment (abrown) [reply crates/runtime/src/memory.rs:459]: Yeah, the `RwLock` could be considered overkill here. Looking at https://github.com/WebAssembly/threads/issues/26 shows some different options for implementing the locking around this. Initially I was considering locking everything (even `memory.size` through a host call) but since the addition of `AtomicUsize` I migrated towards locking `memory.grow` and allowing atomic access to the length. The read locks used in this struct could eventually be replaced by atomic loads and this `RwLock` with a `Mutex` but at the moment the extra safety seems fine and this type of refactor could be resolved in the tracking issue.

File: crates/runtime/src/memory.rs
Comment (cfallin) [reply crates/runtime/src/memory.rs:546]: OK awesome -- just a comment here pointing to that and documenting why it's OK would be great then :-)

File: crates/runtime/src/vmcontext.rs
Comment (abrown) [reply crates/runtime/src/vmcontext.rs:231]: I believe the reason @alexcrichton suggested this relaxed function was for all the cases where the length is retrieved but no synchronization was necessary (i.e., anything related to owned memory). I could certainly make this `SeqCst` but that doesn't totally make sense for all the owned memory cases. Since, like @cfallin mentioned, `current_length()` can only differ by appearing smaller than what another thread sees, I will update the documentation comment accordingly. This function should be precise for any owned memories (regardless of what ordering is used) and will be safely imprecise for shared memories but under-estimating (at least until the spec changes to allow memory shrinking in some way).

File: crates/runtime/src/instance.rs
Comment (abrown) [reply crates/runtime/src/instance.rs:205]: I don't see it. `Instance::memory` is called by `Instance::get_memory` which is used by `memory_copy`, `memory_fill`, etc. That functionality should still be present for shared memory I think.

File: crates/wasmtime/src/memory.rs
Comment (abrown) [reply crates/wasmtime/src/memory.rs]: The `VMMemoryDefinition` is stored on the `wasmtime_runtime::SharedMemory` already; maybe that's what you mean. Storing the `VMMemoryImport` would mean that the shared memory is associated with a single instance in a single store... can't these things be shared across stores?

In any case, this function is only used when importing a shared memory into an instance from `OwnedImports::push` so it seems like the extra work is only done then. If the above question is answered, perhaps this refactor is a suitable candidate for the shared memory tracking issue?

File: crates/runtime/src/instance.rs
Comment (alexcrichton) [reply crates/runtime/src/instance.rs:205]: Sorry I couldn't place this comment well due to the PR not actually covering the relevant line of code, I mean the `set_memory` function just below this comment

File: crates/runtime/src/memory.rs
Comment (alexcrichton) [inline crates/runtime/src/memory.rs:624]: Reading over this again, I think this should return an error that this isn't supported because this won't work correctly with the pooling allocator as the lifetime of the linear memory is tied ot the lifetime of the instance, which isn't correct in the case of shared memories.

File: crates/runtime/src/memory.rs
Comment (alexcrichton) [inline crates/runtime/src/memory.rs:472]: Can these asserts be deferred to `SharedMemory::wrap` and can this method call `SharedMemory::wrap`? Otherwise `Memory::new_dynamic` which I believe is called for locally defined memories isn't asserting these invariants (when it needs to be doing so)

File: crates/runtime/src/vmcontext.rs
Comment (alexcrichton) [reply crates/runtime/src/vmcontext.rs]: One thing to consider here is the effect on the synchronization of this precise field, the current length, but there's also the effects of whether we can see the rest of concurrent memory modifications or such. For example if we read the current length at a particular value what does that mean for reading other pieces of memory?

I'm not sure what the answer is but I'm pretty sure it's affected by the memory orderings here. I don't know what the desired memory orderings are so I would personally use `SeqCst` everywhere because I know it's not wrong. Then again `SeqCst` can also be buggy because it provides too strong of an ordering which means we can't relax it in the future when it turns out we're not supposed to have such a strong ordering.

Anyway these are also theoretical concerns of mine and don't necessarily need any changes at this time.

File: crates/runtime/src/memory.rs
Comment (abrown) [reply crates/runtime/src/memory.rs:589]: Ok, I conditioned the "replace the memory definition" code on whether the memory was shared or not which then makes it possible to make this `unreachable!`. I tried changing `set_memory` to take an `OwnedMemoryIndex` but calculating that is not fast: we have to iterate through all of the memories to find the owned ones. We do this already during instantiation because we're iterating through the memories anyways but it seems expensive to do for every `memory.grow` (we could obviously save off a map of "defined to owned indexes" during instantiation but then that means extra space).

File: crates/runtime/src/instance.rs
Comment (abrown) [reply crates/runtime/src/instance.rs:205]: Ok, see my comment [here](https://github.com/bytecodealliance/wasmtime/pull/4187#discussion_r891694994): it's tricky to re-calculate the owned index for every `memory.grow`.

File: crates/runtime/src/memory.rs
Comment (abrown) [reply crates/runtime/src/memory.rs:467]: No, it is possible that multiple threads could be asking for `byte_size()` and this allows those calls to be concurrent. Eventually we could move to using the `current_length: AtomicUsize` exclusively for `byte_size` but now that I think of it `maximum_byte_size` and `needs_init` still only need read access, not a full lock.

File: crates/runtime/src/memory.rs
Comment (abrown) [reply crates/runtime/src/memory.rs]: But we should still be able to create instances in the pooling allocator that contain `(memory _ _ shared)`, right? I agree that there is something to be fixed here (maybe in the tracking issue?) but it's not clear what that would be--prevent deallocation in the pool for memories once they are shared outside the instance?

Comment (alexcrichton) [reply crates/runtime/src/memory.rs]: In theory that should work yeah but it's currently not sound to do so with the pooling allocator. I think it's fine to return an error and leave a "FIXME" for implementing this in the future. For now though I'm trying to make sure that there aren't any known unsoundnesses in the implementation and this is one that needs to be fixed.

File: tests/all/threads.rs
Comment (alexcrichton) [reply tests/all/threads.rs:261]: Certainly!

File: crates/runtime/src/memory.rs
Comment (alexcrichton) [reply crates/runtime/src/memory.rs]: Nah `todo!` is fine, just something to indicate us to come back to this if anyone stumbles across it.

**Review by alexcrichton — APPROVED** [review]
Ok thanks again for doing all this @abrown and being patient with reviews! I'm going to approve and merge this now. Further work is tracked on https://github.com/bytecodealliance/wasmtime/issues/4245 but I think this is a solid enough base to build on which should be workable enough for the next level of experimentation and feedback on the implementation.

**Pull request merged**
